ntoskrnl/io/iomgr/error.c

```diff
@@ -669,20 +689,29 @@ IoRaiseHardError(IN PIRP Irp,
         return;
     }
 
-    /* Setup an APC */
-    ErrorApc = ExAllocatePoolWithTag(NonPagedPool,
-                                     sizeof(KAPC),
-                                     TAG_APC);
+    // TODO: Do the APC only if this is for an asynchronous operation;
```

Comment:
I wouldn't quite say it that way. The APC is the normal case, the work item is for very special circumstances (cleanup/fsctl IRPs for paging I/O only)

Reply:
The documentation says it always sends an APC and I don't any room for other cases there
https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntddk/nf-ntddk-ioraiseharderror

Reply:
The documentation doesn't say "always"; it talks in generic terms.

Reply:
I've reformulated the comment based on yours from mattermost, please check :)